authui/src/authflowv2/form.ts

```diff
@@ -0,0 +1,144 @@
+import { visit, clearCache } from "@hotwired/turbo";
+import { Controller } from "@hotwired/stimulus";
+import axios, { Method } from "axios";
+import {
+  disableAllButtons,
+  hideProgressBar,
+  showProgressBar,
+  progressEventHandler,
+} from "../loading";
+import { handleAxiosError } from "../messageBar";
+
+// Turbo has builtin support for form submission.
+// We once migrated to use it.
+// However, redirect to external is broken because
+// the redirect is made with fetch, which is subject to CORS.
+// A typical problem is the support for the post login redirect URI.
+// The post login redirect URI is usually an external link that
+// the origin of the link does not list our origin as allowed origin.
+// If we rely on Turbo to handle form submission,
+// the post login redirect URI will be broken.
+// Therefore, we reverted to handle form submission ourselves.
+// To disable the builtin form submission of Turbo,
+// I studied its source code and discovered that
+// Turbo checked in the bubble event listener to see if
+// the event is prevented.
+// So I added a capture event listener to call preventDefault()
+// to stop Turbo from submitting forms.
+//
+// See https://github.com/authgear/authgear-server/issues/2333
+export class XHRSubmitFormController extends Controller {
+  revertDisabledButtons: { (): void } | null = null;
+  forms: HTMLFormElement[] = [];
+
+  onSubmitCapture = (e: Event) => {
+    e.preventDefault();
+  };
+
+  onSubmit = (e: Event) => {
+    this.submitForm(e);
+  };
+
+  async submitForm(e: Event) {
+    const form = e.currentTarget as HTMLFormElement;
+
+    if (form.querySelector('[data-turbo="false"]')) {
+      return;
+    }
+
+    // Do not stop propagation so that GTM can recognize the event as Form Submission trigger.
+    // e.stopPropagation();
+
+    const formData = new FormData(form);
+
+    const params = new URLSearchParams();
+    formData.forEach((value, name) => {
+      params.set(name, value as string);
+    });
+    const loadingEvent = new CustomEvent<{ button?: HTMLButtonElement }>(
+      "onLoading",
+      {
+        detail: {},
+      }
+    );
+    // FormData does not include any submit button's data:
+    // include them manually, since we have at most one submit button per form.
+    const submitButtons = form.querySelectorAll('button[type="submit"]');
+    for (let i = 0; i < submitButtons.length; i++) {
+      const button = submitButtons[i] as HTMLButtonElement;
+      params.set(button.name, button.value);
+    }
+    if (form.id) {
```

Comment:
It should work even the form doesn't have an id. Usually it is unneccesary to define form id in the button.